test/fika/type_checker_test.exs

```diff
@@ -287,6 +287,28 @@ defmodule Fika.TypeCheckerTest do
     end
   end
 
+  describe "map" do
+    test "type check for a valid expression" do
+      str = ~s({"foo" => 123, "bar" => 345})
+
+      ast = TestParser.expression!(str)
+
+      assert {:ok, "Map(String,Int)", _} = TypeChecker.infer_exp(Env.init(), ast)
+    end
+
+    test "type check for map with mixed type" do
+      str = ~s({1 => [1, 2], "foo" => 345})
+
+      ast = TestParser.expression!(str)
+
+      assert {
+               :error,
+               "Elements of map have different types. " <>
+                 "Expected: Map(Int,List(Int)), got: Map(String,Int)"
```

Comment:
We shouldn't have to infer the type of 345 because "foo" is encountered with a wrong type so we can halt early. That means the error message can be "Expected map key of type Int, but got String.".

Reply:
@emilsoman To do this, we'd first check all keys and then move onto validating the value types. What I tried was to consider checking kv pairs and match the differences.
If the concern is spending extra time, then I will change it to checking keys first.
If the concern is the error message, I could split the type matches and give a proper message.
wdyt? 😄

Reply:
We can walk through the list of k_v pairs and as soon as we find a wrong key or wrong value type, we return the appropriate error at the first sight. This means the error will be either "Expected map's key to be .. but got .." or "Expected map's value to be .. but got ..".

Reply:
Another way I can think of is to take first element as key-val pair, then for the rest of the elements only type-check value if key's type-check passes. Is this what you meant?

Reply:
Spoke too soon before refreshing^ 😂 
Think I got it now, will push a change